microprofile/cdi/src/main/java/io/helidon/microprofile/cdi/HelidonProxyServices.java

```diff
@@ -110,39 +133,65 @@ public boolean supportsClassDefining() {
 
         LOGGER.finest("Defining class " + className + " original class: " + originalClass.getName());
 
+        MethodHandles.Lookup lookup;
+
         try {
-            Module classModule = originalClass.getModule();
-            if (!myModule.canRead(classModule)) {
+            // lookup class name "guessed" from the class name of the proxy
+            // proxy name must contain the $ - if it does not, we just use the originalClass as that is safe
+            int index = className.indexOf('$');
+
+            Class<?> lookupClass;
+            if (index < 0) {
+                LOGGER.finest(() -> "Attempt to define a proxy class without a $ in its name. Class name: " + className + ","
+                                      + " original class name: " + originalClass.getName());
+                lookupClass = originalClass;
+            } else {
+                // I would like to create a private lookup in the same package as the proxied class, so let's do it
+                // use the "extracted" lookup class name, if that fails, use the original class
+                lookupClass = tryLoading(originalClass, className.substring(0, className.indexOf('$')));
+            }
+
+            Module lookupClassModule = lookupClass.getModule();
+            if (!myModule.canRead(lookupClassModule)) {
                 // we need to read the module to be able to create a private lookup in it
                 // it also needs to open the package we are doing the lookup in
-                myModule.addReads(classModule);
+                myModule.addReads(lookupClassModule);
             }
+
             // next line would fail if the module does not open its package, with a very meaningful error message
-            MethodHandles.Lookup lookup = MethodHandles.privateLookupIn(originalClass, MethodHandles.lookup());
-            if (classBytes.length == len) {
-                return lookup.defineClass(classBytes);
-            } else {
-                byte[] bytes = new byte[len];
-                System.arraycopy(classBytes, off, bytes, 0, len);
-                return lookup.defineClass(bytes);
-            }
+            lookup = MethodHandles.privateLookupIn(lookupClass, MethodHandles.lookup());
         } catch (IllegalAccessException e) {
             throw new RuntimeException("Failed to define class " + className, e);
         }
+
+        return defineClass(lookup, className, classBytes, off, len);
     }
 
-    private boolean samePackage(Class<?> originalClass, String className) {
-        String origPackage = originalClass.getPackageName();
-        String newPackage = packageName(className);
-        return newPackage.equals(origPackage);
+    private Class<?> tryLoading(Class<?> originalClass, String className) {
+        try {
+            ClassLoader cl = originalClass.getClassLoader();
+            return cl.loadClass(className);
```

Comment:
Nit: just do `return originalClass.getClassLoader().loadClass(className)`; no need to define a local variable.